File: Braille-Assistant/app/src/main/java/org/example/brailleassistant/utils/BrailleCellFinder.java
package org.example.brailleassistant.utils;

import org.opencv.core.Mat;
import org.opencv.core.Point;
import org.opencv.core.Scalar;
import org.opencv.imgproc.Imgproc;

import java.util.ArrayList;
import java.util.Arrays;


public class BrailleCellFinder {

    public class BrailleCellFinderOutput_t {

        public short[] brailleLines;
        public short[] brailleCellColumns;
        public boolean validity;

        public BrailleCellFinderOutput_t(short[] brailleLines, short[] brailleCellColumns) {
            this.brailleLines = brailleLines;
            this.brailleCellColumns = brailleCellColumns;
            this.validity = false;
        }
    }

    private class EmptyRowsAndColumns_t {

        short[] emptyRows;
        short[] emptyColumns;

        private EmptyRowsAndColumns_t(short[] emptyRows, short[] emptyColumns) {
            this.emptyRows = emptyRows;
            this.emptyColumns = emptyColumns;
        }
    }

    public BrailleCellFinder() {

    }

    public BrailleCellFinderOutput_t calculateBrailleCells(Mat brailleImage) { // BrailleCellFinderOutput_t

        EmptyRowsAndColumns_t empty = this.findEmptyRowsAndColumns(brailleImage);
        short[] brailleLines = this.findBrailleLines(empty.emptyRows);
        short[] brailleCellColumns = this.findBrailleCellColumns(empty.emptyColumns);
        BrailleCellFinderOutput_t cellFinderOutput = new BrailleCellFinderOutput_t(brailleLines, brailleCellColumns);
        if (cellFinderOutput.brailleCellColumns == null || cellFinderOutput.brailleLines == null) {

            cellFinderOutput.validity = false;

        } else {

            this.validateIdentifiedCells(cellFinderOutput);

        }

        return cellFinderOutput;

    }

    private void validateIdentifiedCells(BrailleCellFinderOutput_t identifiedBrailleCells) {

        int numberOfBrailleLines = identifiedBrailleCells.brailleLines.length / 2;
        int[] lineWidths = new int[numberOfBrailleLines];

        if (identifiedBrailleCells.brailleLines.length == 2) {
            // do nothing
        } else if (identifiedBrailleCells.brailleLines.length > 2) {
            for (int i = 0; i < identifiedBrailleCells.brailleLines.length; i+=2) {
                lineWidths[i/2] = identifiedBrailleCells.brailleLines[i + 1] - identifiedBrailleCells.brailleLines[i];
            }
            Arrays.sort(lineWidths);
            int medianWidth = lineWidths[numberOfBrailleLines / 2];
            if((lineWidths[0] < 0.75*medianWidth) || (lineWidths[numberOfBrailleLines - 1] > 1.25*medianWidth)) {
                return;
            }

        } else {
            return;
        }

        int numberOfBrailleColumns = identifiedBrailleCells.brailleCellColumns.length / 2;
        int[] cellWidths = new int[numberOfBrailleColumns];

        if (identifiedBrailleCells.brailleCellColumns.length == 2) {
            // do nothing
        } else if (identifiedBrailleCells.brailleCellColumns.length > 2) {
            for (int i = 0; i < identifiedBrailleCells.brailleCellColumns.length; i+=2) {
                cellWidths[i/2] = identifiedBrailleCells.brailleCellColumns[i + 1] - identifiedBrailleCells.brailleCellColumns[i];
            }
            Arrays.sort(cellWidths);
            int medianWidth = cellWidths[numberOfBrailleColumns / 2];
            if((cellWidths[0] < 0.75*medianWidth) || (cellWidths[numberOfBrailleColumns - 1] > 1.25*medianWidth)) {
                return;
            }
        } else {
            return;
        }

        identifiedBrailleCells.validity = true;

    }

    private EmptyRowsAndColumns_t findEmptyRowsAndColumns(Mat brailleImage) {

        short[] emptyRows = new short[brailleImage.height()];
        short numberOfEmptyRows = 0;
        long rowTotal;

        short[] emptyColumns = new short[brailleImage.width()];
        short numberOfEmptyColumns = 0;
        long columnTotal;

        int size = (int) (brailleImage.total() * brailleImage.channels());
        byte[] temp = new byte[size];
        brailleImage.get(0, 0, temp);

        long rowLimit = (brailleImage.width() * (brailleImage.height() - 1));
        int rowIncrement = brailleImage.width();
        short rowNumber = 0;

        // Empty Rows
        for (int j = 0; j <= rowLimit; j += rowIncrement) {

            rowTotal = 0;
            rowNumber++;


            for (short i = 0; i < brailleImage.width(); i++) {

                rowTotal += temp[j + i];


            }

            if (rowTotal == 0) {

                emptyRows[numberOfEmptyRows] = rowNumber;
                numberOfEmptyRows++;

            }

        }

        // Empty Columns
        for (short i = 0; i < brailleImage.width(); i++) {

            columnTotal = 0;



            for (int j = 0; j <= rowLimit; j += rowIncrement) {

                columnTotal += temp[j + i];


            }

            if (columnTotal == 0) {

                emptyColumns[numberOfEmptyColumns] = i;
                numberOfEmptyColumns++;

            }

        }

        return new EmptyRowsAndColumns_t(Arrays.copyOfRange(emptyRows, 0, numberOfEmptyRows), Arrays.copyOfRange(emptyColumns, 0, numberOfEmptyColumns));

    }

    private short[] findBrailleLines(short[] emptyRows) {

        ArrayList<Short> emptyRowLabelsList = new ArrayList<>();

        // Initialise emptyRowLabel
        short[] emptyRowLabel = new short[2];
        emptyRowLabel[0] = emptyRows[0]; // first empty row
        emptyRowLabel[1] = 0; // number of consecutive empty rows

        for (short i = 1; i < emptyRows.length; i++) {
            if ((emptyRows[i] - emptyRows[i - 1]) == 1) { // consecutive empty rows
                emptyRowLabel[1] += 1;
            } else {
                emptyRowLabelsList.add(emptyRowLabel[0]);
                emptyRowLabelsList.add(emptyRowLabel[1]);
                emptyRowLabel[0] = emptyRows[i];
                emptyRowLabel[1] = 0;
            }
        }
        // add last label if there is a sequence of empty rows at the bottom of the image
        if (emptyRowLabel[1] > 0) {
            emptyRowLabelsList.add(emptyRowLabel[0]);
            emptyRowLabelsList.add(emptyRowLabel[1]);
        }

        // convert to array
        Short[] emptyRowLabelsArray = emptyRowLabelsList.toArray(new Short[emptyRowLabelsList.size()]);

        // calculate mean number of consecutive empty rows
        int meanRowGap = 0;
        for (short i = 1; i < emptyRowLabelsArray.length; i+=2) {
            meanRowGap += emptyRowLabelsArray[i];
        }
        meanRowGap = meanRowGap / (emptyRowLabelsArray.length / 2);

        // find braille lines
        ArrayList<Short> emptyRowLabelsOfInterestList = new ArrayList<>();

        for (short i = 0; i < emptyRowLabelsArray.length; i+=2) {
            if (emptyRowLabelsArray[i + 1] > meanRowGap) {
                emptyRowLabelsOfInterestList.add(emptyRowLabelsArray[i]);
                emptyRowLabelsOfInterestList.add(emptyRowLabelsArray[i + 1]);
            }
        }

        // convert to array
        Short[] emptyRowLabelsOfInterestArray = emptyRowLabelsOfInterestList.toArray(new Short[emptyRowLabelsOfInterestList.size()]);

        if (emptyRowLabelsOfInterestArray.length > 1) {
            short[] brailleLines = new short[emptyRowLabelsOfInterestArray.length - 2];

            for (short i = 0; i < emptyRowLabelsOfInterestArray.length - 2; i += 2) {
                brailleLines[i] = (short) (emptyRowLabelsOfInterestArray[i] + emptyRowLabelsOfInterestArray[i + 1]);
                brailleLines[i + 1] = emptyRowLabelsOfInterestArray[i + 2];
            }

            return brailleLines;
        } else {

            return null;

        }
    }

    private short[] findBrailleCellColumns(short[] emptyColumns) {

        ArrayList<Short> emptyColumnLabelsList = new ArrayList<>();

        // Initialise emptyRowLabel
        short[] emptyColumnLabel = new short[2];
        emptyColumnLabel[0] = emptyColumns[0]; // first empty row
        emptyColumnLabel[1] = 0; // number of consecutive empty rows

        for (short i = 1; i < emptyColumns.length; i++) {
            if ((emptyColumns[i] - emptyColumns[i - 1]) == 1) { // consecutive empty rows
                emptyColumnLabel[1] += 1;
            } else {
                emptyColumnLabelsList.add(emptyColumnLabel[0]);
                emptyColumnLabelsList.add(emptyColumnLabel[1]);
                emptyColumnLabel[0] = emptyColumns[i];
                emptyColumnLabel[1] = 0;
            }
        }
        // add last label if there is a sequence of empty rows at the bottom of the image
        if (emptyColumnLabel[1] > 0) {
            emptyColumnLabelsList.add(emptyColumnLabel[0]);
            emptyColumnLabelsList.add(emptyColumnLabel[1]);
        }

        // convert to array
        Short[] emptyColumnLabelsArray = emptyColumnLabelsList.toArray(new Short[emptyColumnLabelsList.size()]);

        // calculate mean number of consecutive empty rows
        int meanColumnGap = 0;
        for (short i = 1; i < emptyColumnLabelsArray.length; i+=2) {
            meanColumnGap += emptyColumnLabelsArray[i];
        }
        meanColumnGap = meanColumnGap / (emptyColumnLabelsArray.length / 2);

        ArrayList<Short> emptyColumnLabelsOfInterestList = new ArrayList<>();

        for (short i = 0; i < emptyColumnLabelsArray.length; i+=2) {
            if (emptyColumnLabelsArray[i + 1] > meanColumnGap) {
                emptyColumnLabelsOfInterestList.add(emptyColumnLabelsArray[i]);
                emptyColumnLabelsOfInterestList.add(emptyColumnLabelsArray[i + 1]);
            }
        }

        // convert to array
        Short[] emptyColumnLabelsOfInterestArray = emptyColumnLabelsOfInterestList.toArray(new Short[emptyColumnLabelsOfInterestList.size()]);

        if (emptyColumnLabelsOfInterestArray.length > 1) {
            short[] brailleCellColumns = new short[emptyColumnLabelsOfInterestArray.length - 2];

            for (short i = 0; i < emptyColumnLabelsOfInterestArray.length - 2; i += 2) {
                 brailleCellColumns[i] = (short) (emptyColumnLabelsOfInterestArray[i] + emptyColumnLabelsOfInterestArray[i+1]);
                 brailleCellColumns[i+1] = (short) (emptyColumnLabelsOfInterestArray[i+2]);
            }

            return brailleCellColumns;

        } else {

            return null;

        }

    }

}

File: Braille-Algorithms/CellFinder/BrailleCellFinder.py
import numpy as np


class BrailleCellFinder(object):

    def __findEmptyRows(self, input_image):
        empty_rows = []
        for i in range(input_image.shape[0]):  # shape[0] = height of image (number of rows)
            if np.sum(input_image[i, :]) == 0:
                empty_rows.append(i)
        return empty_rows

    def __findEmptyColumns(self, input_image):
        empty_columns = []
        column_summation = np.sum(input_image, axis=0)
        for i in range(input_image.shape[1]): # shape[1] = width of image (number of columns)
            if column_summation[i] == 0:
                empty_columns.append(i)
        return empty_columns

    def __findBrailleLines(self, empty_rows):
        # Group consecutive empty rows and label
        empty_row_labels_list = []
        # get first empty row  and set consecutive empty rows to 0
        empty_row_label = [empty_rows[0], 0]
        for i in range(1, len(empty_rows)):  # iterate through empty rows from element 1 to last element
            if empty_rows[i] - empty_rows[i - 1] == 1:  # empty rows are consecutive
                empty_row_label[1] += 1
            else:  # empty rows aren't consecutive
                empty_row_labels_list.append(empty_row_label)  # add empty row label
                empty_row_label = [empty_rows[i], 0]  # initialise next empty row label
        # add last label if there is a sequence of empty rows at bottom of image
        if empty_row_label[1] > 0:
            empty_row_labels_list.append(empty_row_label)

        # Calculate mean number of consecutive empty rows
        empty_row_labels = np.asarray(empty_row_labels_list)
        mean_row_gap = np.mean(empty_row_labels[:, 1])

        # Identify the row labels corresponding to the gaps between braille lines
        row_labels_of_interest = []
        braille_lines = []
        for label in empty_row_labels:
            if label[1] > mean_row_gap:
                row_labels_of_interest.append(label)
        for i in range(0, len(row_labels_of_interest) - 1):
            braille_lines.append(
                [row_labels_of_interest[i][0] + row_labels_of_interest[i][1], row_labels_of_interest[i + 1][0]])
        return braille_lines

    def __findBrailleCellColumns(self, empty_columns):
        # Group consecutive empty columns and label
        empty_column_labels_list = []
        # get first empty column  and set consecutive empty columns to 0
        empty_column_label = [empty_columns[0], 0]
        for i in range(1, len(empty_columns)):  # iterate through empty columns from element 1 to last element
            if empty_columns[i] - empty_columns[i - 1] == 1:  # empty columns are consecutive
                empty_column_label[1] += 1
            else:  # empty columns aren't consecutive
                empty_column_labels_list.append(empty_column_label)  # add empty column label
                empty_column_label = [empty_columns[i], 0]  # initialise next empty column label
        # add last label if there is a sequence of empty columns at bottom of image
        if empty_column_label[1] > 0:
            empty_column_labels_list.append(empty_column_label)

        # Calculate mean number of consecutive empty columns
        empty_column_labels = np.asarray(empty_column_labels_list)
        mean_column_gap = np.mean(empty_column_labels[:, 1])

        # Identify the column labels corresponding to the gaps between braille cells
        column_labels_of_interest = []
        braille_cell_columns = []
        for label in empty_column_labels:
            if label[1] > mean_column_gap / 1.3:
                column_labels_of_interest.append(label)
        for i in range(0, len(column_labels_of_interest) - 1):
            braille_cell_columns.append([column_labels_of_interest[i][0] + column_labels_of_interest[i][1],
                                         column_labels_of_interest[i + 1][0]])
        return braille_cell_columns

    def __validateIdentifiedCells(self, braille_lines, braille_cell_columns):
        line_widths = []
        cell_widths = []

        if len(braille_lines) > 0:
            for line in braille_lines:
                line_widths.append(line[1]-line[0])
            mean_line_width = sum(line_widths) / len(line_widths)
            for width in line_widths:
                if (width < 0.85*mean_line_width) or (width > 1.15*mean_line_width):
                    return False

        if len(braille_cell_columns) > 0:
            for cell in braille_cell_columns:
                cell_widths.append(cell[1]-cell[0])
            mean_cell_width = sum(cell_widths) / len(cell_widths)
            for width in cell_widths:
                if (width < 0.8*mean_cell_width) or (width > 1.2*mean_cell_width):
                    return False

        return True



    def calculateBrailleCells(self, input_image):
        empty_rows = self.__findEmptyRows(input_image)
        empty_columns = self.__findEmptyColumns(input_image)
        braille_lines = self.__findBrailleLines(empty_rows)
        braille_cell_columns = self.__findBrailleCellColumns(empty_columns)
        validation = self.__validateIdentifiedCells(braille_lines, braille_cell_columns)
        print(braille_lines)
        print("\n \n \n")
        print(braille_cell_columns)
        return validation, braille_lines, braille_cell_columns


